taiga_halo2/src/note.rs

```diff
@@ -155,14 +147,37 @@ impl Note {
             note_type,
             app_data_dynamic,
             value,
-            nk_container,
+            nk_container: NullifierKeyContainer::Key(nk),
             is_merkle_checked,
             psi: rseed.get_psi(&rho),
             rcm: rseed.get_rcm(&rho),
             rho,
         }
     }
 
+    // The rho, psi, and rcm are not specified until the action is constructed.
+    #[allow(clippy::too_many_arguments)]
+    pub fn new_output_note(
```

Comment:
I did it deliberately.
I want the user to ignore the pair relation of input and output, so I push back the output reset_rho until the action construction and reset it automatically.
If we allow passing the corresponding input note to create the output, the user must know the pair relation and remember the order of the notes when constructing the action circuits.
Does it make sense?

Reply:
Hm. Yes and no... It makes sense to help the user not set rho manually, but if the user supplies an output note and then taiga resets rho, which would (I assume?) change the commitment of the note, then the user might not realize that the commitment has changed.

Reply:
The `rho` acts as the nonce of note. It's meaningless to specify the `rho` of the output note before constructing the action unless the user knows the action pair beforehand. But it's annoying to assume the user has the prior knowledge.
IMP, the `rho` should be set by the protocol, not the user. It might not be a reset_rho, but a set_rho to be more accurate. The user only supplies the necessary info of the output note(like value, type, static_data, etc.); the action/protocol supplies other information (like the rho).
If we don't want the user to specify a default `rho` implicitly, we could adjust the `rho` format to the `Option<Nullifier>` and set a `None` for output notes.

Reply:
What if we go deeper here? From the user point of view, the only note fields that are important are `note_type`, `app_data_dynamic`, `value`, and maybe `is_merkle_checked`. Everything else is quite internal. What do you think if we split the note into two portions: a user facing part, and a full internal note that contains the user facing part plus the internal fields. Then the user only ever needs to think about the actual contents of the note.

Reply:
Yeah, that's what I want to fulfill ultimately. When the Anoma spec comes out, we can have some refactoring on the Note struct if needed.
For now, I only used some proper Note APIs to let users ignore the internal parameters.

Reply:
I'm a little confused by the new API. Let's say, I have a note `input1` and I want to create a ptx to swap it to a different note. Which APIs would I call to prepare the 4 notes for the ptx?

Reply:
Do we need to supply the `nk_com`? Can't it be generated by Taiga?

Reply:
In general, the `nk_com` should be provided by the receiver/owner along with the `app_data_dynamic.`
If it is an intent note, the creator can generate a random `nk,` get the `nk_com` to fill the output note, and expose the `nk` to the solver(or someone who will consume the intent).

Reply:
Right, right. I keep forgetting how nullifier derivation works.